NBitcoin.Tests/OutputDescriptorTests.cs

```diff
@@ -0,0 +1,348 @@
+		{
+			var keysPriv = new FlatSigningRepository();
+			var keysPub = new FlatSigningRepository();
+			var leftPath = pathIndex;
+
+			var parsePriv = OutputDescriptor.Parse(MaybeUseHInsteadOfApostrophe(priv), false, keysPriv);
+			var parsePub = OutputDescriptor.Parse(MaybeUseHInsteadOfApostrophe(pub), false, keysPub);
+
+			// Check that correct output type is inferred
+			Assert.Equal(parsePriv.GetScriptPubKeyType(), spkType);
+			Assert.Equal(parsePub.GetScriptPubKeyType(), spkType);
+
+			// 1. Check private keys are extracted from the private version but not the public one.
+			Assert.True(keysPriv.Secrets.Count > 0);
+			Assert.True(keysPub.Secrets.Count == 0);
+
+			// 2. Check both will serialize back to the public version.
+			string pub1 = parsePriv.ToString();
+			string pub2 = parsePub.ToString();
+			Assert.True(EqualDescriptorStr(pub, pub1));
+			Assert.True(EqualDescriptorStr(pub, pub2));
+
+			// 3. Check that both can be serialized with private key back to the private version, but not without private key.
+			Assert.True(parsePub.TryGetPrivateString(keysPriv, out var priv1));
+			Assert.True(EqualDescriptorStr(priv, priv1));
+			Assert.False(parsePriv.TryGetPrivateString(keysPub, out priv1));
+			Assert.True(parsePub.TryGetPrivateString(keysPriv, out priv1));
+			Assert.True(EqualDescriptorStr(priv, priv1));
+			Assert.False(parsePub.TryGetPrivateString(keysPub, out priv1));
+
+			// Check that `IsRange()` on both returns the expected result.
+			Assert.Equal(parsePub.IsRange(), (flags & RANGE) != 0);
+			Assert.Equal(parsePriv.IsRange(), (flags & RANGE) != 0);
+
+			// * For ranged descriptors,  the `scripts` parameter is a list of expected result outputs, for subsequent
+			//   positions to evaluate the descriptors on (so the first element of `scripts` is for evaluating the
+			//   descriptor at 0; the second at 1; and so on). To verify this, we evaluate the descriptors once for
+			//   each element in `scripts`.
+			// * For non-ranged descriptors, we evaluate the descriptors at positions 0, 1, and 2, but expect the
+			//   same result in each case, namely the first element of `scripts`. Because of that, the size of
+			//   `scripts` must be one in that case
+			bool isRange = (flags & RANGE) > 0;
+			if (!isRange) Assert.Single(scripts);
+			var max = isRange ? scripts.Length : 3;
+
+			for (int i = 0; i < max; ++i)
+			{
+				var expectedScript = scripts[isRange ? i : 0];
+				for (int t = 0; t < 2; ++t)
+				{
+					bool isHardend = (flags & HARDENED) != 0;
+					var keyProvider = isHardend ? keysPriv : keysPub;
+
+					var scriptProvider = new FlatSigningRepository();
+					Assert.True((t != 0 ? parsePriv : parsePub).TryExpand((uint)i, keyProvider.GetPrivateKey, scriptProvider, out var spks));
+					Assert.Equal(spks.Count, expectedScript.Length);
+
+					// --- cache ---
```

Comment:
If you want to move your hand while reviewing, it may be good idea to port the caching part from bitcoin-core. the test is in [Line 148](https://github.com/bitcoin/bitcoin/blob/6a2ba626852bd5bd8531dfa335c78fa87e4f6282/src/test/descriptor_tests.cpp#L148)-215 . Which should fit in this line.
@lontivero